File: backend/app/services/day_service.py
"""Day service."""

from datetime import date
from typing import List, Optional

from sqlalchemy.orm import Session

from app.models.day import Day


class DayService:
    """Service for day operations."""

    @staticmethod
    def get_or_create_day(db: Session, user_id: int, day_date: date) -> Day:
        """Get existing day or create a new one.

        Args:
            db: Database session
            user_id: User ID
            day_date: Date for the day

        Returns:
            Day object (existing or newly created)
        """
        # Try to get existing day
        day = (
            db.query(Day)
            .filter(Day.user_id == user_id, Day.date == day_date)
            .first()
        )

        if day:
            return day

        # Create new day
        new_day = Day(user_id=user_id, date=day_date)
        db.add(new_day)
        db.commit()
        db.refresh(new_day)
        return new_day

    @staticmethod
    def get_day(db: Session, user_id: int, day_date: date) -> Optional[Day]:
        """Get specific day for user.

        Args:
            db: Database session
            user_id: User ID
            day_date: Date for the day

        Returns:
            Day object or None if not found
        """
        return (
            db.query(Day)
            .filter(Day.user_id == user_id, Day.date == day_date)
            .first()
        )

    @staticmethod
    def get_days_range(
        db: Session, user_id: int, start_date: date, end_date: date
    ) -> List[Day]:
        """Get all days in date range for user.

        Args:
            db: Database session
            user_id: User ID
            start_date: Start date (inclusive)
            end_date: End date (inclusive)

        Returns:
            List of Day objects ordered by date
        """
        return (
            db.query(Day)
            .filter(
                Day.user_id == user_id,
                Day.date >= start_date,
                Day.date <= end_date,
            )
            .order_by(Day.date)
            .all()
        )

    @staticmethod
    def update_day(db: Session, day_id: int, **kwargs) -> Day:
        """Update day fields.

        Args:
            db: Database session
            day_id: Day ID
            **kwargs: Fields to update (tag, feeling, effort_score, summary, llm_advice)

        Returns:
            Updated Day object

        Raises:
            ValueError: If day not found
        """
        day = db.query(Day).filter(Day.id == day_id).first()

        if not day:
            raise ValueError(f"Day with id {day_id} not found")

        # Update allowed fields
        allowed_fields = {"tag", "feeling", "effort_score", "summary", "llm_advice"}
        for field, value in kwargs.items():
            if field in allowed_fields:
                setattr(day, field, value)

        db.commit()
        db.refresh(day)
        return day

    @staticmethod
    def delete_day(db: Session, day_id: int) -> bool:
        """Delete day.

        Args:
            db: Database session
            day_id: Day ID

        Returns:
            True if deleted, False if not found
        """
        day = db.query(Day).filter(Day.id == day_id).first()

        if not day:
            return False

        db.delete(day)
        db.commit()
        return True

File: mobile/src/services/api/dayService.ts
import { apiClient } from './apiClient';
import type { Day } from '../../types/models/health';

export const dayService = {
  /**
   * Get days within date range
   */
  getDays: async (startDate?: string, endDate?: string): Promise<Day[]> => {
    const params = new URLSearchParams();
    if (startDate) params.append('start_date', startDate);
    if (endDate) params.append('end_date', endDate);

    const response = await apiClient.get(`/days?${params.toString()}`);
    return response.data;
  },

  /**
   * Get single day by ID
   */
  getDay: async (dayId: number): Promise<Day> => {
    const response = await apiClient.get(`/days/id/${dayId}`);
    return response.data;
  },

  /**
   * Get day by date (YYYY-MM-DD)
   */
  getDayByDate: async (date: string): Promise<Day | null> => {
    try {
      const response = await apiClient.get(`/days/date/${date}`);
      return response.data;
    } catch (error: any) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  },

  /**
   * Create new day
   */
  createDay: async (date: string): Promise<Day> => {
    const response = await apiClient.post('/days', { date });
    return response.data;
  },

  /**
   * Update day fields (tag, feeling, effort_score, weight, summary)
   */
  updateDay: async (
    dayId: number,
    data: Partial<Pick<Day, 'tag' | 'feeling' | 'effort_score' | 'weight' | 'summary'>>
  ): Promise<Day> => {
    const response = await apiClient.put(`/days/${dayId}`, data);
    return response.data;
  },

  /**
   * Delete day
   */
  deleteDay: async (dayId: number): Promise<void> => {
    await apiClient.delete(`/days/${dayId}`);
  },
};
